### PR TITLE
feat(executor): added MsgUpsertSequencer consensus message

### PR DESCRIPTION
**Pull request by keruch**
# PR Standards

ADR link https://www.notion.so/dymension/ADR-x-Sequencer-Reward-Address-On-Rollapp-da84af6888c141d0a4c1a8df256a5025
Closes https://github.com/dymensionxyz/dymension/issues/1248

---

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

File: block/produce.go [hunk 1]
```diff
@@ -148,14 +159,24 @@ func (m *Manager) produceBlock(allowEmpty bool, nextProposerHash *[32]byte) (*ty
...
+		proposerHashForBlock = nextProposerInfo.nextProposerHash
+		nextProposerAddr = nextProposerInfo.nextProposerAddr
+		lastProposerBlock = true
 	}
-	block = m.Executor.CreateBlock(newHeight, lastCommit, lastHeaderHash, proposerHashForBlock, m.State, maxBlockDataSize)
+	consensusMsgs, err := m.consensusMsgsOnCreateBlock(nextProposerAddr, lastProposerBlock)
+	if err != nil {
+		return nil, nil, fmt.Errorf("create consensus msgs for create block: last proposer block: %v, height: %d, next proposer addr: %s: %w: %w", lastProposerBlock, newHeight, nextProposerAddr, err, ErrNonRecoverable)
+	}
+	block = m.Executor.CreateBlock(newHeight, lastCommit, lastHeaderHash, proposerHashForBlock, m.State, maxBlockDataSize, consensusMsgs...)
```
Comment (jgimeno) [inline block/produce.go:179]: Ideally there should be only one point of adding consensus messages. Given they come from the ConsensusMessagesStream this should send them there.

The only problem I see here is that there is no implementation of the stream as of now. I would put a message somewhere with a TODO to unify it.

**Review by omritoptix — APPROVED** [review]
looks good. small code cleanliness comments.

**Review by omritoptix — CHANGES_REQUESTED** [review]
I notice now we introduce a circular dependency with the RDK which makes this pr fails to build also.

I think we should avoid the circular depedency, it makes it really difiicult to maintain after. 

can't we just copy the protos?

File: p2p/client_test.go [hunk 2]
```diff
@@ -49,6 +50,7 @@ func TestClientStartup(t *testing.T) {
 	assert.NoError(err)
 }
 
+// TestBootstrapping TODO: this test is flaky in main. Try running it 100 times to reproduce.
```
Comment (mtsitrin) [inline p2p/client_test.go:53]: add reference to https://github.com/dymensionxyz/dymint/issues/869

File: block/produce.go [hunk 3]
```diff
@@ -96,18 +100,26 @@ func (m *Manager) ProduceBlockLoop(ctx context.Context, bytesProducedC chan int)
 	}
 }
 
+// nextProposerInfo holds information about the next proposer.
+type nextProposerInfo struct {
```
Comment (mtsitrin) [inline block/produce.go:104]: this `nextProposerInfo struct` is uneeded, and changes alot of APIs just to be used in `consensusMsgsOnCreateBlock`.
`nextProposerAddr` can be queried using `m.State.Sequencers.GetByHash` (similar to `SetByHash`)

File: block/produce.go [hunk 4]
```diff
@@ -171,6 +197,45 @@ func (m *Manager) produceBlock(allowEmpty bool, nextProposerHash *[32]byte) (*ty
 	return block, commit, nil
 }
 
+// consensusMsgsOnCreateBlock forms a list of consensus messages that need execution on rollapp's BeginBlock.
+// Currently, we need to create a sequencer in the rollapp if it doesn't exist in the following cases:
+//   - On the very first block after the genesis or
+//   - On the last block of the current sequencer (eg, during the rotation).
+func (m *Manager) consensusMsgsOnCreateBlock(
```
Comment (mtsitrin) [inline block/produce.go:204]: i suggest to move to separate file, to have all consesnsMsg related logic

File: block/produce.go [hunk 5]
```diff
@@ -6,18 +6,22 @@ import (
...
-	"github.com/dymensionxyz/dymint/store"
-	uevent "github.com/dymensionxyz/dymint/utils/event"
-
+	"github.com/gogo/protobuf/proto"
 	tmed25519 "github.com/tendermint/tendermint/crypto/ed25519"
 	cmtproto "github.com/tendermint/tendermint/proto/tendermint/types"
 	tmtypes "github.com/tendermint/tendermint/types"
 	tmtime "github.com/tendermint/tendermint/types/time"
 
+	sequencertypes "github.com/dymensionxyz/dymint/types/pb/rollapp/sequencers/types"
```
Comment (mtsitrin) [inline block/produce.go:18]:
```suggestion
	rdktypes "github.com/dymensionxyz/dymint/types/pb/rollapp/sequencers/types"
```

File: block/produce.go [hunk 6]
```diff
@@ -171,6 +197,45 @@ func (m *Manager) produceBlock(allowEmpty bool, nextProposerHash *[32]byte) (*ty
...
+	if err != nil {
+		return nil, fmt.Errorf("convert tendermint pubkey to cosmos: %w", err)
+	}
+	anyPK, err := codectypes.NewAnyWithValue(pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("convert cosmos pubkey to any: %w", err)
+	}
+
+	return []proto.Message{&sequencertypes.ConsensusMsgUpsertSequencer{
+		Operator:   nextProposerSettlementAddr,
```
Comment (mtsitrin) [inline block/produce.go:233]: we're passing "dym1...." to the RDK, which needs to be converted to the rollapp's native prefix.
Maybe we should pass the []byte of the address, not the bech32 representation

**Review by mtsitrin — CHANGES_REQUESTED** [review]

**Pull request closed without merging**